### PR TITLE
Move the subscriptions empty state screen from WooPayments to core

### PR DESCRIPTION
**Pull request by james-allan**
Fixes #https://github.com/woocommerce/woocommerce-subscriptions/issues/4549

## Description

In WooPayments v3.4 we introduced empty state screens for the WooCommerce Subscriptions admin list table. see https://github.com/Automattic/woocommerce-payments/pull/3403

As part of that PR there were 2 variations of that screen. 

1. If the merchant had not onboarded to WooPayments it would include a [button for them to onboard](https://user-images.githubusercontent.com/8490476/142810806-73e70740-03bb-4d32-a108-5ffc86def5a9.png). 
2. If the merchant had onboarded, it would include base screen which mentioned that subscriptions will be listed here and a link to create a product. See screenshot below.

**This PR moves the base screen to subscriptions core so that all subscription-core client products (WooPayments and Woo Subscriptions) have the updated empty state screen.**

Screenshot 2023-08-02 at 9 46 56 am

> **Note**
> This will require a set of accompanying changes in Woo Payments that this version of core is merged into to ensure no breakage. 

## How to test this PR

> **Warning**
> When testing this PR beware that the function used to determine whether to display the empty state (ie do subscriptions exist) uses a caching system and so it's suggested that if you flip between having and not having subscriptions that you clear the cache or bypass it in `cache_and_get()`.

1. Make sure you have WC Subscriptions plugin active.
2. Delete all your subscriptions.
   - _If you'd prefer to keep them, you can do a bulk post type update and restore it after testing._
3. Go to the **WooCommerce > Subscriptions** screen. 
   - On trunk you'll see an empty list with simple html elements (p and a tags).
   - On this branch you should see the new empty state screen.
4. The button should take you to the edit product page.
5. Purchase a subscription. 
6. Go to the subscriptions admin list table.
    - You should see the subscription you just purchased listed.
7. Filter the list table by searching for an dummy string - we want to test 0 results.
    - You should just see the "No subscriptions found." text.

This PR deprecates the old hook `woocommerce_subscriptions_not_found_label`. This hook used to filter the plain text "No Subscriptions found", however it was being used by WooPayments to also return full HTML content. Given the name ("label"), it no longer makes sense for it to exist in that form. I've deprecated it, still call it but have replaced it with a new hook `woocommerce_subscriptions_not_found_html`.

To test the deprecation you can enable WooPayments, disable WC Subscriptions and reload the WooCommerce > Subscriptions screen. You should see the same empty state but this deprecation warning should be shown.

```
woocommerce_subscriptions_not_found_label is deprecated since version 6.1.0! Use woocommerce_subscriptions_not_found_html instead.
```

## Product impact

- [x] Added changelog entry (or does not apply)
- [ ] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [ ] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [ ]  Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

### CODE REVIEW

File: assets/css/admin-empty-state.css [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+.woo_subscriptions_empty_state__container .woo_subscriptions_empty_state__description {
+	max-width: 535px;
+	margin: 15px 0;
+}
+.woo_subscriptions_empty_state__button_container {
+	margin: 8px 0;
+}
+.woo_subscriptions_empty_state__container .woo_subscriptions_empty_state__button_container * + * {
+	margin-left: 10px;
+}
```
Comment (james-allan) [inline assets/css/admin-empty-state.css:26]: These styles were copied from the WooPayments file. Found here: https://github.com/Automattic/woocommerce-payments/blob/6.0.0/client/subscriptions-empty-state/style.scss. 

Note I haven't copied over the TOS styling as we don't need that in the base empty state screen and I've converted that `scss` into plain `css`.

File: includes/class-wc-subscriptions-core-plugin.php [hunk 2]
```diff
@@ -326,19 +327,6 @@ public function get_gateways_handler_class() {
...
-			$not_found_text = __( 'No Subscriptions found', 'woocommerce-subscriptions' );
-		} else {
-			$not_found_text = '<p>' . __( 'Subscriptions will appear here for you to view and manage once purchased by a customer.', 'woocommerce-subscriptions' ) . '</p>';
-			// translators: placeholders are opening and closing link tags
-			$not_found_text .= '<p>' . sprintf( __( '%1$sLearn more about managing subscriptions &raquo;%2$s', 'woocommerce-subscriptions' ), '<a href="http://docs.woocommerce.com/document/subscriptions/store-manager-guide/#section-3" target="_blank">', '</a>' ) . '</p>';
-			// translators: placeholders are opening and closing link tags
-			$not_found_text .= '<p>' . sprintf( __( '%1$sAdd a subscription product &raquo;%2$s', 'woocommerce-subscriptions' ), '<a href="' . esc_url( WC_Subscriptions_Admin::add_subscription_url() ) . '">', '</a>' ) . '</p>';
-		}
-
-		$subscriptions_not_found_text = apply_filters( 'woocommerce_subscriptions_not_found_label', $not_found_text );
```
Comment (james-allan) [inline includes/class-wc-subscriptions-core-plugin.php:339]: All this logic was kinda hard to understand the purpose of tbh. It looks as though there's a filter in here that I'll need to honour though. 

If subscriptions do exists (`$subscriptions_exist === true`) or a third-party says they exists via the `woocommerce_subscriptions_not_empty` filter, but WP doesn't load anything into the table it use to say "No subscriptions found".

Screenshot 2023-08-02 at 10 50 22 am

Whereas if there are no subscriptions or something doesn't change the `$subscriptions_exist` value via the filter, then it displays this:

Screenshot 2023-08-02 at 11 32 53 am

I don't know if we want to preserve this nuance at all.

Comment (james-allan) [reply includes/class-wc-subscriptions-core-plugin.php:339]: wait 🤦‍♂️. 

This is used when you use filters an no subscriptions are returned. 🤔

Comment (james-allan) [reply includes/class-wc-subscriptions-core-plugin.php:339]: I've fixed these issues in more recent commits.

File: templates/admin/html-admin-empty-list-table.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * The template for displaying the empty state for the subscriptions list table.
+ *
+ * @version 6.1.0
```
Comment (mattallan) [inline templates/admin/html-admin-empty-list-table.php]:
```suggestion
 * @version 6.2.0
```

Comment (mattallan) [reply templates/admin/html-admin-empty-list-table.php]: Small typo 😄

File: includes/admin/class-wcs-admin-empty-list-content-manager.php [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+			$content = wc_get_template_html(
+				'html-admin-empty-list-table.php',
+				[],
+				'',
+				WC_Subscriptions_Core_Plugin::instance()->get_subscriptions_core_directory( 'templates/admin/' )
+			);
+		}
+
+		// Backwards compatibility for the woocommerce_subscriptions_not_found_label filter.
+		if ( has_action( 'woocommerce_subscriptions_not_found_label' ) ) {
```
Comment (mattallan) [inline includes/admin/class-wcs-admin-empty-list-content-manager.php:41]: As you recently found, we could use the following code here:

```
$content = apply_filters_deprecated( 'woocommerce_subscriptions_not_found_label', [ $content ], '6.2.0', 'woocommerce_subscriptions_not_found_content' );
```

But I think the current code is fine and consistent with what we've done in other areas

**Review by mattallan — APPROVED** [review]
The changes LGTM :100: 

I tested the following flows:
- [x] On a store with no subscriptions (new empty state)
- [x] On a store with subscriptions (regular list table)
- [x] On a store with subscriptions filtering the table to have results ("No subscriptions found.")

---

While testing this on HPOS stores, I noticed the new empty state is never being used and at first I thought it was because `'edit-shop_subscription' !== $screen->id` doesn't work witj HPOS, but no... it doesn't work because you can't actually change the default "No items found." message.

Here's the Subscriptions table on a fresh store with HPOS enabled:

image

I had a look into how we could update this to use this empty state for HPOS environments and it's a bit complicated, most probably impossible without changes to WC Core.

We would need update the [`ListTable`](https://github.com/woocommerce/woocommerce/blob/dda93bfe51bbbf706f877ef6d84b111eb5438dec/plugins/woocommerce/src/Internal/Admin/Orders/ListTable.php#L15) class in WC Core which extends `WP_List_Table` and update it to override the [`no_items()` function](https://github.com/WordPress/wordpress-develop/blob/6.3/src/wp-admin/includes/class-wp-list-table.php#L338) and have it return whatever is set as "not_found" arg when registering the order type.

Anyway, bit of a rabbit hole, but this is good to be merged.

**Pull request merged**